jni/src/faiss_wrapper.cpp

```diff
@@ -223,28 +223,16 @@ jobjectArray knn_jni::faiss_wrapper::QueryIndex_WithFilter(knn_jni::JNIUtilInter
     omp_set_num_threads(1);
     // create the filterSearch params if the filterIdsJ is not a null pointer
     if(filterIdsJ != nullptr) {
-        int *filteredIdsArray = jniUtil->GetIntArrayElements(env, filterIdsJ, nullptr);
+        long *filteredIdsArray = jniUtil->GetLongArrayElements(env, filterIdsJ, nullptr);
         int filterIdsLength = env->GetArrayLength(filterIdsJ);
         std::unique_ptr<faiss::IDSelector> idSelector;
-        FilterIdsSelectorType idSelectorType = getIdSelectorType(filteredIdsArray, filterIdsLength);
-        // start with empty vectors for 2 different types of empty Selectors. We need define them here to avoid copying of data
-        // during the returns. We could have used pass by reference, but we choose pointers. Returning reference to local
-        // vector is also an option which can be efficient than copying during returns but it requires upto date C++ compilers.
-        // To avoid all those confusions, its better to work with pointers here. Ref: https://cplusplus.com/forum/general/56177/
-        std::vector<faiss::idx_t> convertedIds;
-        std::vector<uint8_t> bitmap;
-        // Choose a selector which suits best
-        if(idSelectorType == BATCH) {
-            convertedIds.resize(filterIdsLength);
-            convertFilterIdsToFaissIdType(filteredIdsArray, filterIdsLength, convertedIds.data());
-            idSelector.reset(new faiss::IDSelectorBatch(convertedIds.size(), convertedIds.data()));
+        if(filterIdsTypeJ == BITMAP) {
+            const int bitsetArraySize = filterIdsLength * 8;
+            uint8_t *bitmap = reinterpret_cast<uint8_t*>(filteredIdsArray);
```

Comment:
@heemin32  Would you pls tell me what your concern details about the safety?

As i see:
In query we reused JAVA element like
```
float* rawQueryvector = jniUtil->GetFloatArrayElements(env, queryVectorJ, nullptr);
indexReader->search(1, rawQueryvector, kJ, dis.data(), ids.data());
```

AND i checked OpenJDK code:

https://github.com/openjdk/jdk/blob/5a74c2a67ebcb47e51732f03c4be694bdf920469/src/hotspot/share/prims/jni.cpp#L2411-L2443 

it shows that when we calling `GetLongArrayElements/GetIntArrayElements`, it would do memory copy from jvm to native memory. so when we called `long *filteredIdsArray = jniUtil->GetLongArrayElements(env, filterIdsJ, nullptr);`  it returned address as same as c++ code malloc memory. so we must care about the memory overflow with length which is `bitsetArraySize`. 

AND even we using a vector<uint8_t> like before, we still need care about memory safety which `convertedIds.data()` is a pointer address: 

```
idSelector.reset(new faiss::IDSelectorBatch(convertedIds.size(), convertedIds.data()));
```

Reply:
cpp `long` can be 32bits in some compiler or platform. Should we use `int64_t` instead?

Reply:
> cpp `long` can be 32bits in some compiler or platform. Shouldn't we use `int64_t` instead?

@heemin32 Thanks for your reply. it is nice! LGTM for this. fixed it at https://github.com/opensearch-project/k-NN/pull/1402/commits/c2cd334b9b9bbf56379fff728fe703b6e618b4ea